### ISSUE TITLE
Organizational grouping field types (Category Group, Sections, etc)

### ISSUE THREAD

**Issue by Mosnar** [craftcms/cms#7762]
### Description

It would be nice if Craft provided some fields to help build element queries based on CMS structures. Ideally, I could set them in single-select mode (dropdown) or multi-select mode (checkboxes?) It's important to be able to restrict selection to specific items or "All" items in the field settings.

Here are some ideas for the field types this might accommodate:

- User Groups
- Asset Volumes
- Asset Transforms
- Category Groups
- Entry Sections
- Entry Types
- Tag Groups
- Commerce: Product Types
- Commerce: Sales
- Commerce: Discounts
- Commerce: Shipping Methods

I think there are a few different ways to implement this, but generally speaking, I think these are the requirements:
- Relationships should be defined by a uid to ensure it plays nicely with project-config
- The value of the field should be an instance(s) of the model representing the data (e.g. ProductType)
- Multi-select or single-select mode
- Select from "All" or pre-defined subset of options

**Comment by wedot** [comment]
+1 🎉

**Comment by brandonkelly** [comment]
Can you provide an example use case, and how this would work?

**Comment by Mosnar** [comment]
@brandonkelly Sure - here are some example hypothetical scenarios:

- I want to add a field to allow authors to restrict certain entries to specific user groups (imagine membership tiers)
- I have sections for "Blog", "News", and "Case Studies" - the listing pages all use the same template and entry type, with just a dropdown to pick which section to pull content from
- I have a matrix block in a content builder for embedding images and I want to let the author pick an Asset Transform to use
- I have a widget that can appear on my homepage to show a random assortment of products - but I only want to pull specific product types. A "Product Type" field could be added to help me when I inevitably forget to add new product types to this dropdown.
- I have a landing page that will pull in all products that belong to a Craft Commerce sale I selected from a dropdown
- I wrote a custom integration that pushes new Assets that are uploaded to a virus scanner. I could drop an "Asset Volume" field into a global set to act as the UI for controlling which volumes my integration should run on.

Obviously some of these field types are less useful than others (e.g. tag groups) but I wanted to include them to be thorough. 

I think providing individual fields for each of these is probably not the best approach. A couple approaches that come to mind are:
1. Introduce a new option to dropdown and checkbox fields for selecting a custom data source that could be pre-configured somewhere else, such as a php config file. This could work similarly to Redactor & Element API configs. Imagine you have a file in your config folder: `config/field-data/states.php` that returns an array of states or `config/field-data/usergroups.php` returns an array of user groups.
2. Refactor `BaseOptionsField` to trigger an event to override options instead - then we could just use a module to provide the data. Not as elegant, but less complex and more versatile.
